spec/controllers/repositories_controller_spec.rb

```diff
@@ -29,15 +29,64 @@
 require 'spec_helper'
 
 describe RepositoriesController, :type => :controller do
-  let(:project) { FactoryGirl.create(:project) }
-  let(:user) { FactoryGirl.create(:user, :member_in_project => project,
-                                          :member_through_role => role) }
-  let(:repository) { FactoryGirl.create(:repository, :project => project) }
+  let(:project) do
+    project = FactoryGirl.create(:project)
+    allow(Project).to receive(:find).and_return(project)
+    project
+  end
 
-  before do
+  let(:user) { FactoryGirl.create(:user, member_in_project: project,
+                                         member_through_role: role) }
+  let(:repository) do
     allow(Setting).to receive(:enabled_scm).and_return(['Filesystem'])
-    repository  # ensure repository is created after stubbing the setting
+    repo = FactoryGirl.build_stubbed(:repository,
+                                     project: project)
+    allow(repo).to receive(:default_branch).and_return('master')
+    allow(repo).to receive(:branches).and_return(['master'])
+
+    repo
+  end
+
+  before do
     allow(User).to receive(:current).and_return(user)
+    allow(project).to receive(:repository).and_return(repository)
+  end
+
+  describe '#edit' do
+    let(:role) { FactoryGirl.create(:role, permissions: [:manage_repository]) }
+
+    before do
+      # authorization checked in spec/permissions/manage_repositories_spec.rb
+      allow(controller).to receive(:authorize).and_return(true)
+    end
+
+    shared_examples_for "successful response" do
+      it 'is successful' do
+        expect(response).to be_success
+      end
+
+      it 'renders the template' do
+        expect(response).to render_template 'projects/settings/repository'
+      end
+    end
+
+    context 'GET' do
+      before do
+        xhr :get, :edit
+      end
+
+      it_behaves_like "successful response"
+    end
+
+    context 'POST' do
+      before do
+        allow(repository).to receive(:save).and_return(true)
+
+        xhr :post, :edit
+      end
+
+      it_behaves_like "successful response"
```

Comment:
Prefer single-quoted strings when you don't need string interpolation or special symbols.